### PR TITLE
lightning: save chunk checkpoint timely

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve? 
Try to save chunk checkpoint timely.

### What is changed and how it works?
- Return a chunk status object when close chunk but it's not flushed. Then before each chunk restorer started, check and save all the pending chunk checkpoint if available.
- Try to save the chunk checkpoint before and after each deliver loop.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes
 - Has interface methods change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Release Note

 - check and save chunk checkpoint timely.

### CODE REVIEW

File: pkg/lightning/backend/local/local.go [hunk 1]
```diff
@@ -165,6 +165,13 @@ type File struct {
 	sstIngester    sstIngester
 	finishedRanges syncedRanges
 
+	// sst seq lock
+	seqLock sync.Mutex
+	// seq number for incoming sst meta
+	nextSeq atomic.Int32
```
Comment (kennytm) [inline pkg/lightning/backend/local/local.go:171]: seems unnecessary to make `nextSeq` atomic as it is only used inside `e.seqLock`

Comment (glorv) [reply pkg/lightning/backend/local/local.go]: yes, fixed.

File: pkg/lightning/backend/importer/importer_test.go [hunk 2]
```diff
@@ -115,8 +115,9 @@ func (s *importerSuite) TestWriteRows(c *C) {
 	c.Assert(err, IsNil)
 	err = writer.WriteRows(s.ctx, nil, s.kvPairs)
 	c.Assert(err, IsNil)
-	err = writer.Close(s.ctx)
+	st, err := writer.Close(s.ctx)
```
[the remark this answers is not in the document]
Comment (glorv) [reply pkg/lightning/backend/importer/importer_test.go:118]: checked in L120, though it should be nil for importer backend

File: pkg/lightning/backend/local/local.go [hunk 3]
```diff
@@ -2519,23 +2550,37 @@ func (w *Writer) flush(ctx context.Context) error {
...
 	// FIXME: in theory this line is useless, but In our benchmark with go1.15
 	// this can resolve the memory consistently increasing issue.
 	// maybe this is a bug related to go GC mechanism.
 	w.writeBatch = nil
-	return err
+	return flushStatus{local: w.local, seq: w.lastMetaSeq}, err
+}
+
+func (w *Writer) IsSynced() bool {
+	return w.batchCount == 0 && w.lastMetaSeq <= w.local.finishedMetaSeq.Load()
```
Comment (Little-Wallace) [inline pkg/lightning/backend/local/local.go:2583]: Does it need to be called in a mutex?

**Review by Little-Wallace — COMMENTED** [review]
rest LGTM

**Review by Little-Wallace — APPROVED** [review]
LGTM

**Comment by gozssky** [comment]
LGTM

**Review by gozssky — APPROVED** [review]

**Review by kennytm — APPROVED** [review]

**Pull request merged**